### ISSUE TITLE
Numerically stable summation methods

### ISSUE THREAD

**Issue by ibab** [tensorflow/tensorflow#2625]
It would be nice to be able to use numerically stable summation methods like
- pairwise summation
- Kahan summation

for applications where numerical accuracy is important.
This could work with an optional `stable` keyword argument to `tf.reduce_sum` or via a `tf.stable_reduce_sum` op.

I think implementing Kahan summation in the Eigen tensor library wouldn't be difficult at all, we would simply have to add a stateful `KahanSumReducer`.
It should also be possible to provide a vectorized version.
Pairwise summation might be more difficult, as the reduction code would have to be touched.

**Comment by suiyuan2009** [comment]
dibs

**Comment by ibab** [comment]
@suiyuan2009: Cool, but you should get some input from one of the tensorflow devs before starting. @benoitsteiner, would this be a reasonable contribution?

**Comment by egpbos** [comment]
Any news? I could make good use of this.

**Comment by suiyuan2009** [comment]
@egpbos want to contribute?

**Comment by egpbos** [comment]
In principle, yes, but I'm still deciding whether to go with TensorFlow in our project in the first place, which most importantly depends on #5323. Might take a few more weeks before we decide.

**Comment by girving** [comment]
Looks like this fell through the cracks.  Good idea though, so I'll leave it open.

**Comment by lakshayg** [comment]
@girving I am trying to get this into the Eigen library but the mailing list/issue tracker seem very inactive and I think that getting it into Eigen is going to take a long time. Is there some way I could implement this within tensorflow?

**Comment by girving** [comment]
@rmlarsen Any suggestions for how to do Kahan summation without changing Eigen?  One problem is that an ideal implementation would probably stride the Kahan summation algorithm based on the architecture optimal packet size.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by evolu8** [comment]
TF members, I'm not sure I understand the tensorflowbutler comments, but it seems the issue creator is not responding to the request to drop requests for contributions. Can you fix this your end? It would suck for this issue to stall due to miscommunication/poor timing of interests. It is now 2 years old, but very important to those who care about it.

**Comment by cheungdaven** [comment]
Encounter the same problem here, the differences are huge for my task.  Any replacements?

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by jpreiss** [comment]
Is there any update on this issue? I think it's especially important for policy gradient reinforcement learning, since the basin of attraction is less stable. I can run the same RL algorithm twice with all other sources of nondeterminism eliminated, and get significantly different results each time. I would like to publish reproducible results, even if the performance hit is significant.

**Comment by ppwwyyxx** [comment]
This snippet might be related:
```python
import tensorflow as tf
import numpy as np
data = np.random.rand(int(1e4), int(1e4)) + np.random.rand(int(1e4), 1)
data = data.astype('float32')
vec = tf.placeholder(tf.float32, data.shape)
s = tf.reduce_sum(vec)
with tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=1)) as sess:
    np_sum = np.sum(data)
    tf_sum = sess.run(s, feed_dict={vec: data})
    diff = abs(tf_sum - np_sum)
    print(diff, diff/1e8)   # difference can be as large as 200656, when running with CPU.

data = data.astype('float64')
vec = tf.placeholder(tf.float64, data.shape)
s = tf.reduce_sum(vec)
with tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=1)) as sess:
    np_sum = np.sum(data)
    tf_sum = sess.run(s, feed_dict={vec: data})
    diff = abs(tf_sum - np_sum)
    print(diff, diff/1e8)
```

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by ekelsen** [comment]
unassign

**Comment by gonnet** [comment]
Would it be sufficient to su recursively across the dimensions of a Tensor?

This would reduce the problem from being O(size) to O(size of largest dimension), which might just be enough for most use-cases, and should be a lot easier to implement than compensated summation.

**Comment by bhack** [comment]
@tensorflow/dev-support Can you remove the contribution welcome label and retriage this?

I think we don't have a clear enough contribution path to label this as contribution welcome so we need to re-triage it. 

This ticket is very old but recently we had another related bug at https://github.com/tensorflow/tensorflow/issues/58133

**Comment by cantonios** [comment]
Eigen already does tree reduction (similar to pairwise summation) on CPU for floating-point types (https://gitlab.com/libeigen/eigen/-/blob/master/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h#L240).

On GPU, it ends up being (mostly) pairwise: https://gitlab.com/libeigen/eigen/-/blob/master/unsupported/Eigen/CXX11/src/Tensor/TensorReductionGpu.h#L541

We do not currently do kahan summation in addition to tree reduction.  Adding it to TensorFlow would involve implementing it in Eigen for CPU/GPU (potentially with vectorization).  Someone can raise a feature request over there: https://gitlab.com/libeigen/eigen/-/issues/new

**Comment by sushreebarsa** [comment]
Hi,
Kindly refer to the comment above.
Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow version with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the release notes to stay up to date with the latest developments which are happening in the Tensorflow space.

Thank you!
